pkg/api/validate/openshiftcluster_validatedynamic.go

```diff
@@ -51,30 +61,81 @@ func NewOpenShiftClusterDynamicValidator(log *logrus.Entry, env env.Interface, o
 	}
 }
 
-type azureClaim struct {
-	Roles []string `json:"roles,omitempty"`
-}
-
-func (*azureClaim) Valid() error {
-	return fmt.Errorf("unimplemented")
+// NewOpenShiftClusterSlimDynamicValidator creates a new OpenShiftClusterSlimDynamicValidator
+func NewOpenShiftClusterSlimDynamicValidator(log *logrus.Entry, clientID string, clientSecret api.SecureString, subscriptionID string, tenantID string, clusterSpec *arooperatorapi.ClusterSpec) OpenShiftClusterSlimDynamicValidator {
+	spp := &api.ServicePrincipalProfile{
+		ClientID:     clientID,
+		ClientSecret: clientSecret,
+	}
+	return &openShiftClusterSlimDynamicValidator{
+		log:            log,
+		spp:            spp,
+		tenantID:       tenantID,
+		subscriptionID: subscriptionID,
+		clusterSpec:    clusterSpec,
+	}
 }
 
-type openShiftClusterDynamicValidator struct {
-	log *logrus.Entry
+// openShiftClusterFullDynamicValidator is dynamic validator used inside RP context
+// and can operate within cluster and firstParty service principals.
+// It includes openShiftClusterSlimDynamicValidator
+type openShiftClusterFullDynamicValidator struct {
+	*openShiftClusterSlimDynamicValidator
 	env env.Interface
 
 	oc              *api.OpenShiftCluster
 	subscriptionDoc *api.SubscriptionDocument
 	fpAuthorizer    refreshable.Authorizer
 
-	fpPermissions     authorization.PermissionsClient
+	fpPermissions authorization.PermissionsClient
+}
+
+// openShiftClusterSlimDynamicValidator is dynamic validator used inside clusters
+// context and can operate ONLY with cluster service principal scope
+type openShiftClusterSlimDynamicValidator struct {
+	log *logrus.Entry
+
+	spp             *api.ServicePrincipalProfile
```

Comment:
can we avoid using `api` in the operator-related code? 
Also as far as I understand it doesn't run only in cluster context - on Create this code will run before the cluster creation, is this understanding correct?

Reply:
This is internal representatio of API. And we already use it with VMSize.
But I think you are right. Gonna change this :)